src/materials/tdycharacteristiccurves.c

```diff
@@ -531,16 +546,18 @@ static PetscErrorCode CubicPolynomialSetup(PetscReal x1, PetscReal x2,
 /// @param [out] *f      function evaluated at value
 /// @param [out] *df_dx  derivative of the function evaluated at value = x
 ///
-///  f(x) = a0 + a1 * x + a2 * x^2 + a2 * x^3
-///  df_dx = a1 + 2 * a2 * x + 3 * a3 * x^2
+///  f(x)  = a0 + a1 * (x-x0)/dx + a2 * ((x-x0)/dx)^2 + a2 * ((x-x0)/dx)^3
+///  df_dx =     a1/dx + 2 * a2 * ((x-x0)/dx)/dx + 3 * a3 * ((x-x0)/dx)^2/dx
 ///
-static PetscErrorCode CubicPolynomialEvaluate(PetscReal *coeffs, PetscReal x,
+static PetscErrorCode CubicPolynomialEvaluate(PetscReal *coeffs, PetscReal x, PetscReal x0, PetscReal dx,
                                               PetscReal *f, PetscReal *df_dx) {
 
   PetscFunctionBegin;
 
-  *f = coeffs[0] + coeffs[1]*x + coeffs[2]*PetscPowReal(x,2.0) + coeffs[3]*PetscPowReal(x,3.0);
-  *df_dx = coeffs[1] + 2.0*coeffs[2]*x + 3.0*coeffs[3]*PetscPowReal(x,2.0);
+  PetscReal xt = (x-x0)/dx;
+
+  *f = coeffs[0] + coeffs[1]*xt + coeffs[2]*PetscPowReal(xt,2.0) + coeffs[3]*PetscPowReal(xt,3.0);
+  *df_dx = coeffs[1]/dx + 2.0*coeffs[2]*xt/dx + 3.0*coeffs[3]*PetscPowReal(xt,2.0)/dx;
```

Comment:
A bit more efficient to use Horner's rule
```suggestion
  *f = coeffs[0] + xt * (coeffs[1] + xt * (coeffs[2] + xt * coeffs[3]));
  *df_dx = (coeffs[1] + xt * (2.0*coeffs[2] + xt * 3.0 * coeffs[3])) / dx;
```